File: lilac/src/test/java/com/hartwig/hmftools/lilac/fragment/NucleotideGeneEnrichmentTest.java
package com.hartwig.hmftools.lilac.fragment;

import static com.hartwig.hmftools.lilac.LilacConstants.A_EXON_BOUNDARIES;
import static com.hartwig.hmftools.lilac.LilacConstants.B_EXON_BOUNDARIES;
import static com.hartwig.hmftools.lilac.LilacConstants.C_EXON_BOUNDARIES;
import static com.hartwig.hmftools.lilac.LilacConstants.HLA_A;
import static com.hartwig.hmftools.lilac.LilacConstants.HLA_B;
import static com.hartwig.hmftools.lilac.LilacConstants.HLA_C;
import static com.hartwig.hmftools.lilac.LilacUtils.namesMatch;
import static com.hartwig.hmftools.lilac.fragment.FragmentUtils.expandIndices;

import static junit.framework.TestCase.assertTrue;

import java.util.List;
import java.util.Set;

import com.google.common.collect.Lists;
import com.google.common.collect.Sets;

import org.junit.Test;

public class NucleotideGeneEnrichmentTest
{
    @Test
    public void testGeneEnrichment()
    {
        NucleotideGeneEnrichment enricher = new NucleotideGeneEnrichment(A_EXON_BOUNDARIES, B_EXON_BOUNDARIES, C_EXON_BOUNDARIES);

        List<Integer> indices = Lists.newArrayList();
        indices.add(337);
        indices.add(348);
        assertGene(enricher, Sets.newHashSet(HLA_A, HLA_B), HLA_A, indices);

        assertGene(enricher, Sets.newHashSet(HLA_A, HLA_B), HLA_B, indices);
        assertGene(enricher, Sets.newHashSet(HLA_C), HLA_C, indices);

        indices.add(365);
        assertGene(enricher, Sets.newHashSet("HLA-A"), "HLA-A", indices);
        assertGene(enricher, Sets.newHashSet("HLA-B"), "HLA-B", indices);
    }

    private void assertGene(
            NucleotideGeneEnrichment enricher,
            Set<String> expectedGenes, String alignedGene, List<Integer> aminoAcideIndices)
    {
        Fragment fragment = create(alignedGene, expandIndices(aminoAcideIndices));
        Fragment result = enricher.enrich(fragment);
        assertTrue(namesMatch(result.getGenes(), expectedGenes));
    }

    private Fragment create(final String gene, final List<Integer> indices)
    {
        List<Integer> qualities = Lists.newArrayList();
        qualities.add(0);
        return new Fragment("id", "", gene, Sets.newHashSet(gene), indices, qualities, Lists.newArrayList("G"));
    }

}

File: lilac/src/test/kotlin/com/hartwig/hmftools/lilackt/nuc/NucleotideGeneEnrichmentTest.kt
package com.hartwig.hmftools.lilackt.nuc

import junit.framework.Assert.assertTrue
import org.junit.Test

class NucleotideGeneEnrichmentTest {

    private val aProteinExonBoundaries = setOf(24, 114, 206, 298, 337, 348, 364, 365)
    private val bProteinExonBoundaries = setOf(24, 114, 206, 298, 337, 348, 362)
    private val cProteinExonBoundaries = setOf(24, 114, 206, 298, 338, 349, 365, 366)
    private val enrichment = NucleotideGeneEnrichment(aProteinExonBoundaries, bProteinExonBoundaries, cProteinExonBoundaries)

    @Test
    fun testGeneEnrichment() {
        assertGene(setOf("HLA-A", "HLA-B"), "HLA-A", listOf(337, 348))
        assertGene(setOf("HLA-A", "HLA-B"), "HLA-B", listOf(337, 348))
        assertGene(setOf("HLA-C"), "HLA-C", listOf(337, 348))

        assertGene(setOf("HLA-A"), "HLA-A", listOf(337, 348, 362))
        assertGene(setOf("HLA-B"), "HLA-B", listOf(337, 348, 362))
    }

    fun assertGene(expectedGenes: Set<String>, alignedGene: String, aminoAcideIndices: List<Int>) {
        val victim = create(alignedGene, aminoAcideIndices.flatMap { listOf(it * 3, it * 3 + 1, it * 3 + 2) })
        val result = enrichment.enrich(listOf(victim))[0]
        assertTrue(result.genes == expectedGenes)
    }

    fun create(gene: String, indices: List<Int>): NucleotideFragment {
        return NucleotideFragment("id", setOf(gene), indices, indices.map { 0 }, indices.map { "G" })
    }


}